dj/api/auth.py

```diff
@@ -0,0 +1,58 @@
+"""
+Auth dependencies
+"""
+import logging
+from http import HTTPStatus
+
+from fastapi import APIRouter, Depends, HTTPException, Request, Security
+from fastapi.security import OAuth2PasswordBearer, OAuth2PasswordRequestForm
+
+from dj.models.user import Token, User
+
+_logger = logging.getLogger(__name__)
+router = APIRouter()
+
+oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")
+
+
+async def get_current_user(
+    token: str = Security(oauth2_scheme),  # pylint: disable=unused-argument
+) -> User:  # pylint: disable=unused-argument
+    """
+    Overrideable placeholder function for getting a user from a token
+    """
+    return User(username="admin")
+
+
+class PermissionsChecker:  # pylint: disable=too-few-public-methods
+    """
+    Overrideable placeholder for checking if a user has required permissions
+    """
+
+    def __init__(self, request: Request):
```

Comment:
Summarizing here from the conversation. Maybe some more dedicated parameters like the verb, the base route, the nodes - more targeted arguments overall to avoid any complexity of handling parsing stuff from the url

Reply:
Good point. We definitely need to land on a design where a custom implementation of a permissions checker doesn't need to have a deep understanding on how to parse a request but I'm wondering if we should go even further than just providing the targeted arguments (since they'll still need to make sense of how to compose the arguments together).

DJ can define the `__init__` logic here in a universal way that doesn't need to be replaced and then a custom implementation just needs to worry about the `__call__` logic. Right now DJ will automatically inject the `self.permission` attribute which would be something like `GET:/nodes/default.foo` and the permissions checker could just use that as a standard permission string. You brought up a great point earlier that for the more complicated sql and data endpoint it's not as straightforward. For example `/sql/?metrics=default.foo&metrics=default.bar&dimensions=default.baz&dimensions=default.qux` doesn't make sense as a single resource identifier.

So maybe we leave it fully in the hands of the OSS logic to determine the array of resource identifiers explicitly for every permission check. For the sql endpoint example, DJ would provide an array that looks like `["GET:/nodes/default.foo", "GET:/nodes/default.bar", "GET:/nodes/default.baz", "GET:/nodes/default.qux"]`. The implementer won't need to know about how those resource identifiers are constructed, just that the user needs to have all of those permissions.

Reply:
Yeah, I'm in favor of that logic living in OSS, and the only part that's specific to each internal implementation is just the part that checks a specific grant and returns `access granted` or `access denied`. This also lets us layer grants better - for example, we could have something like:
1. Collect all nodes and see if they have grants. Apply each node's grants and deny access to the endpoint if any of them fail.
2. If any node doesn't have a grant, we default to the node's namespace's grant and apply that if it exists. If nothing exists, we default to granting access (or denying access).